modules/System Admin/dataRetentionProcess.php

```diff
@@ -0,0 +1,115 @@
+
+You should have received a copy of the GNU General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use Gibbon\Services\Format;
+use Gibbon\Domain\ScrubbableGateway;
+use Gibbon\Domain\System\SettingGateway;
+use Gibbon\Domain\System\DataRetentionGateway;
+
+include '../../gibbon.php';
+
+$URL = $_SESSION[$guid]['absoluteURL'].'/index.php?q=/modules/'.getModuleName($_POST['address']).'/dataRetention.php';
+
+if (isActionAccessible($guid, $connection2, '/modules/System Admin/dataRetention.php') == false) {
+    $URL .= '&return=error0';
+    header("Location: {$URL}");
+} else {
+    $data = [
+        'date' => (!empty($_POST['date'])) ? Format::dateConvert($_POST['date']) : null,
+        'domains' => $_POST['domains'] ?? []
+    ];
+
+    // Validate Inputs
+    if ($data['date'] == '' OR count($data['domains']) < 1) {
+        $URL .= '&return=error3';
+        header("Location: {$URL}");
+        return;
+    } 
+
+    // Ensure table selection persists
+    $container->get(SettingGateway::class)->updateSettingByScope('System', 'dataRetentionDomains', implode(',', $data['domains']));
+
+    // Prepare data retention gateway
+    $dataRetentionGateway = $container->get(DataRetentionGateway::class);
+
+    $allDomains = $dataRetentionGateway->getDomains();
+    $selectedDomains = array_filter($allDomains, function ($key) use (&$data) {
+        return in_array($key, $data['domains']);
+    }, ARRAY_FILTER_USE_KEY);
+
+    $partialFail = false;
+    $scrubbedList = [];
+
+    // Validate each of the domains is scrubbable before proceeding
+    foreach ($selectedDomains as $domain) {
+        foreach ($domain['gateways'] as $gatewayClass) {
+            $gateway = $container->get($gatewayClass);
+            if (!$gateway instanceof ScrubbableGateway) {
+                $URL .= '&return=error2';
+                header("Location: {$URL}");
+                exit;
+            }
+        }
+    }
+
+    // Cycle through each selected domain
+    foreach ($selectedDomains as $domain) {
+        if (empty($domain['gateways'])) continue;
+
+        // Cycle through each gateway and scrub data
+        foreach ($domain['gateways'] as $gatewayClass) {
+            $gateway = $container->get($gatewayClass);
+
+            $scrubbed = $gateway->scrub($data['date'], $domain['context'] ?? []);
+            $scrubbedList = array_merge_recursive($scrubbedList, $scrubbed);
+            $partialFail &= !empty($scrubbed);
+        }
+    }
+
+    // Store a record of which users and which tables were scrubbed
+    foreach ($scrubbedList as $gibbonPersonID => $tables) {
+        if (empty($gibbonPersonID)) continue;
+
+        $data = [
+            'gibbonPersonID'            => $gibbonPersonID,
+            'tables'                    => json_encode(array_values(array_unique($tables))),
+            'status'                    => !empty($tables) ? 'Success' : 'Partial Failure',
```

Comment:
Not sure about the logic here. Can you explain?

Reply:
Sure, this sets up the data to insert into the gibbonDataRetention table, and if there's already a record, it merges the list of tables that have been cleared. I figured we didn't need a timestamp for every time the data has been cleared, just the most recent time, and a record of everything cleared up until that point.

Reply:
Sorry, should have been more specific, looking at line 92. It seems to set Success if there are any tables, otherwise Partial Failure. What happens if some fail and some succeed, would $tables not be empty?